src/storage/heap_file.c

```diff
@@ -7545,8 +7570,24 @@ heap_prepare_get_context (THREAD_ENTRY * thread_p, HEAP_GET_CONTEXT * context, b
 	      assert (false);
 	      goto error;
 	    }
+
+	  ret = pgbuf_check_page_ahead_of_replication (thread_p, context->fwd_page_watcher.pgptr);
+	  if (ret != NO_ERROR)
+	    {
+	      goto error;
+	    }
 	  return S_SUCCESS;
 	}
+      else if (ret == ER_PB_BAD_PAGEID)
+	{
+	  VPID *vpid = pgbuf_get_vpid_ptr (context->fwd_page_watcher.pgptr);
+	  if (vpid == NULL)
+	    {
+	      ret = ER_PB_BAD_PAGEID;
+	      goto error;
+	    }
+	  ret = pgbuf_check_for_deallocated_page_or_desyncronization (thread_p, context->latch_mode, *vpid);
```

Comment:
It seems that if `ER_PB_BAD_PAGEID` occurs because of other cause than the desynchronization, `pgbuf_check_for_deallocated_page_or_desyncronization` returns NO_ERROR, which ends up skipping the error. What do you think?

Reply:
Or is it guaranteed that the error from pgbuf_fix() with OLD_PAGE_DEALLOCATED is always the same as that with OLD_PAGE when it's not from the desynchronization?